README
        __
 .----.|  |--..-----..--.--.
 |  __||  _  ||  _  ||_   _|
 |____||_____||   __||__.__|  connection buffering proxy
------------- |__| ----------------------------------------

Warning note:

 This is experimental software.
 Make sure you know what you're doing before using it for anything else than testing it.

Description:

 cbpx allows you to switch application (clients) between two backends, but only when
 all connections to previously active backend have been closed.

Requirements:

 * Python >=2.6 <3.0

Command line options:

 -p PORT, --port=PORT : port that proxy will listen on
 -a ACTIVE, --active=ACTIVE : IP:port pair of active backend (the one we switch from)
 -s STANDBY, --standby=STANDBY : IP:port pair of standby backend (the one we switch to)
 -r RC_PORT, --rc_port=RC_PORT : port for remote control connections (remote control is disabled if not specified)
 -t SWITCH_MAX_TIME, --switch_max_time=SWITCH_MAX_TIME : timeout (in seconds) after which switchover fails
 -c MAX_QUEUED_CONNS, --max_queued_conns=MAX_QUEUED_CONNS : queued connections limit, after which switchover fails
 -o MAX_OPEN_CONNS, --max_open_conns=MAX_OPEN_CONNS : open connections limit (used for throttling, 0 disables this feature)
 -b LISTEN_BACKLOG, --listen_backlog=LISTEN_BACKLOG :  backlog for listen()
 -n NET_BUFFER_SIZE, --net_buffer_size=NET_BUFFER_SIZE : network communication buffer size
 -w SWITCH_LOOP_WAIT, --switch_loop_wait=SWITCH_LOOP_WAIT : wait in switch loop
 -f LOG_FILE, --log_file=LOG_FILE : log file name
 -l LOG_LEVEL, --log_level=LOG_LEVEL : log level: DEBUG, INFO, WARNING, ERROR
 -x SWITCH_SCRIPT, --switch_script=SWITCH_SCRIPT : script to execute before switch completion (optional)

Proxy Commands:

 hello   : Server greeting (for checking proxy availability in remote control mode)
 help    : Print help 
 quit    : Quit proxy (use 'force' parameter to close all active and ignore queued connections)
 set     : Show/set variables (set PARAMETER VALUE) 
 stats   : Print current statistics (stats [SLEEP] [COUNT]) 
 switch  : Perform the switch
 threads : List alive threads



ChangeLog
0.4.7

* GPL-ed
* scream with error if not all data is sent
* switch script added to 'set' listong options 
* total enqueued connections added back to stats
* shutdown() and close() moved outside the lock in cbpx_transporter.remove()

0.4.6

* 5 connectors as a workaround for lost packets during TCP handshake

0.4.5

* added RelayGate, reworked locking on relay
* backends moved to RelayGate

0.4.4

* transporter loop optimized by removing few calls and one unnecessary check
* switch exec script added
* stats: removed avgcps, added dequeued connections per second
* stats: able to use better resolution
* stats: removed enqueued/dequeued
* removed conn_lock lock

0.4.3

* fixed call to Queue.qsize()

0.4.2

* dead connection removing moved outside the epoll() loop
* release tool #-es l.debug lines for performance ;-)

0.4.1 (experimental)

* single transporter thread, epoll()
* main threads moved to cbpx class, cleaned main script
* log levels fixed: no rubbish at INFO level

0.4.0 (experimental)

* (internal tests: select(), poll(), epoll(), no transporters, single transporter)

0.3.3

* statistics moved to separate class
* connections per second counted
* 'stats' in loop mode checks if arguments are ok

0.3.2

* unknown command doesn't hang RC anymore
* quit ignores options other than 'force'

0.3.1

* 'quit force' added (closes all active connections in SOCK_FUCK_YOU mode)
* 'stats X Y' and 'stats X' sanitized to 'stats' in remote control mode
* 'set' now checks value type and ranges

0.3.0

* fixed log levels: INFO doesn't flood log anymore, only DEBUG does (INFO is default)
* remote control mode doesn't print anything on screen
* print_cfg() moved to cmd.py, print_logo() moved to readline ui
* rewritten switch procedure: used switch timer, new logic, new switch finish reason storage

0.2.10

* cleanup imports (for easier dependency handling)
* rename cmd_interface_ classes to ui_, separate module for ui_*
* python 2.5 enforced in shebang
* 'stats' now behaves even more like vmstat (stats [sleep] [count])
* test_connection() removed (it's not proxy's job)

0.2.9

* log level changes (important messages are INFO/WARNING now)
* TCP remote control
* command interface reworked for remote control
* connector started before listener
* git import, automated versioning (based on git branch), release tool

0.2.9-pre

* quick initial remote control for Karol's tests

0.2.8

* additional code comments
* Event.is_set() -> Event.isSet(): python 2.5 compatibility

0.2.7

* feature: throttling added
* max_open_conns added for throttling
* cleanup: max connection parameters naming fixed
* bugfix: switch_delay fixed

0.2.6

* porządki w kodzie
* nowe parametry wywołania
* switch_delay jest znów honorowany
* przesunięcie sprawdzania przekroczenia max_conn wcześniej w kodzie
* przygotowania do zdalnego sterowania (nie z linii poleceń)

0.2.5

* dodany switch
* feature: switch działa w dwie strony
* bugfix: obsługa ctrl-c przy switchu (przerywa procedurę)
* bugfix: switch zadziała natychmiast, jeśli w momencie uruchamiania go nie ma żadnych aktywnych połączeń
* kolejka ma nieskończoną długość, ale przy wkładaniu sprawdzamy, czy skolejkowanych połączeń nie jest więcej niż max_conn. Jeśli tak, to switch jest przerywany i kolejka spuszczana.
* poziom logowania domyślnie "WARNING" - ponieważ logging jest thread-aware, to robi w chuj locków i psuje wydajność, jak się okazuje.

0.2.4

* development version

0.2.3

* connector_threads usunięte z kodu (spada wydajność)

0.2.2

* connector_threads option displayed in set()

0.2.1

* connector_threads

0.2.0

* queue-only version (no switch)

0.1

* first revision


th.py
#  Copyright (c) 2011-2012 Jakub Filipowicz <jakubf@gmail.com>
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software
#  Foundation, Inc.,
#  51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

from threading import Condition, Lock

# --------------------------------------------------------------------
# Relay gate (based on python's _Event)
class RelayGate():

    # ----------------------------------------------------------------
    def __init__(self):
        self.__cond = Condition(Lock())
        self.__flag = False
        self.reason = "[no reason]"
        self.backends = [('0.0.0.0', 0), ('0.0.0.0', 0)]
        self.backend = 0

    # ----------------------------------------------------------------
    def isSet(self):
        return self.__flag

    # ----------------------------------------------------------------
    def set_backends(self, ai, ap, si, sp):
        self.__cond.acquire()
        try:
            self.backends = [(ai, ap), (si, sp)]
        finally:
            self.__cond.release()

    # ----------------------------------------------------------------
    def switch_backend(self, reason):
        self.__cond.acquire()
        try:
            if not self.__flag:
                self.backend = int(not self.backend)
                self.reason = reason
                self.__flag = True
            self.__cond.notify_all()
        finally:
            self.__cond.release()

    # ----------------------------------------------------------------
    def set(self, reason):
        self.__cond.acquire()
        try:
            if not self.__flag:
                self.__flag = True
                self.reason = reason
            self.__cond.notify_all()
        finally:
            self.__cond.release()

    # ----------------------------------------------------------------
    def clear(self, reason):
        self.__cond.acquire()
        try:
            if self.__flag:
                self.reason = reason
                self.__flag = False
        finally:
            self.__cond.release()

    # ----------------------------------------------------------------
    def wait(self, timeout=None):
        self.__cond.acquire()
        try:
            if not self.__flag:
                self.__cond.wait(timeout)
            return self.__flag
        finally:
            self.__cond.release()

    # ----------------------------------------------------------------
    def get_active(self):
        return self.backends[self.backend]

    # ----------------------------------------------------------------
    def get_standby(self):
        return self.backends[int(not self.backend)]

    # ----------------------------------------------------------------
    def get_reason(self):
        self.__cond.acquire()
        try:
            reason = self.reason
            self.reason = "[no reason]"
        finally:
            self.__cond.release()
        return reason


# vim: tabstop=4 expandtab shiftwidth=4 softtabstop=4


ui.py
#  Copyright (c) 2011-2012 Jakub Filipowicz <jakubf@gmail.com>
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software
#  Foundation, Inc.,
#  51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

import time
import readline
import logging

from socket import *

from utils import params, l
from utils import __version__

# ------------------------------------------------------------------------
class ui:

    # --------------------------------------------------------------------
    def __init__(self):
        pass

    # --------------------------------------------------------------------
    def shutdown(self):
        pass

    # --------------------------------------------------------------------
    def read(self):
        time.sleep(1)
        return "hello"

    # --------------------------------------------------------------------
    def write(self, text):
        pass

    # --------------------------------------------------------------------
    def finish(self):
        pass


# ------------------------------------------------------------------------
class ui_readline(ui):

    # --------------------------------------------------------------------
    def print_logo(self):
        print """        __
 .----.|  |--..-----..--.--.
 |  __||  _  ||  _  ||_   _|
 |____||_____||   __||__.__| %s : connection buffering proxy
------------- |__| -----------------------------------------------
""" % __version__

    # --------------------------------------------------------------------
    def __init__(self):
        l.debug("Starting readline user interface")
        self.print_logo()
        print " Ready for your commands, my master.\n"

    # --------------------------------------------------------------------
    def read(self):
        line = raw_input("cbpx> ")
        return line

    # --------------------------------------------------------------------
    def write(self, text):
        print text


# ------------------------------------------------------------------------
class ui_net(ui):

    # --------------------------------------------------------------------
    def __init__(self):
        l.debug("Starting TCP user interface on port %i" % int(params.rc_port))
        self.rc_sock = socket(AF_INET, SOCK_STREAM)
        self.rc_sock.setsockopt(SOL_SOCKET, SO_REUSEADDR, 1)
        self.rc_sock.bind(('', int(params.rc_port)))
        self.rc_sock.listen(5)

    # --------------------------------------------------------------------
    def shutdown(self):
        l.debug("Shutting down TCP user interface")
        self.rc_sock.shutdown(SHUT_RDWR)
        self.rc_sock.close()

    # --------------------------------------------------------------------
    def read(self):
        l.debug("Awaiting TCP command connection...")
        (self.rc_conn, self.rc_addr) = self.rc_sock.accept()
        l.debug("TCP commnd connection from: %s" % str(self.rc_addr))
        l.debug("Awaiting network command...")
        line = self.rc_conn.recv(int(params.net_buffer_size))
        return self.sanitize(line)

    # --------------------------------------------------------------------
    def sanitize(self, command):
        sanitized = command

        # allow only one-time stats
        if command.startswith("stats") and command != "stats":
            l.debug("Sanitizing command: '%s'" % command)
            sanitized = "stats"

        return sanitized

    # --------------------------------------------------------------------
    def write(self, text):
        try:
            self.rc_conn.send(text + "\n")
        except:
            # we couldn't send command output. so what?
            pass

    # --------------------------------------------------------------------
    def finish(self):
        try:
            self.rc_conn.shutdown(SHUT_RDWR)
            self.rc_conn.close()
        except:
            # so what?
            pass



# vim: tabstop=8 expandtab shiftwidth=4 softtabstop=4


cmd.py
#  Copyright (c) 2011-2012 Jakub Filipowicz <jakubf@gmail.com>
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software
#  Foundation, Inc.,
#  51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

import time
import readline
import logging
import threading

from network import cbpx_connector, cbpx_listener, cbpx_transporter, conn_q, relay, kill_script
from utils import params, l
from utils import __version__
from stats import cbpx_stats


# ------------------------------------------------------------------------
# Timer that handles switch timeout
class SwitchTimer(threading._Timer):

    # --------------------------------------------------------------------
    def __init__(self, timeout):
        threading._Timer.__init__(self, timeout, self.timeout_action)
        self.setName("SWTimer")

    # --------------------------------------------------------------------
    def cancel(self):
        l.debug("canceling switch timer")
        threading._Timer.cancel(self)
        kill_script()

    # --------------------------------------------------------------------
    def timeout_action(self):
        relay.set("switch timeout")
        kill_script()
        l.debug("Switch timer done")


# ------------------------------------------------------------------------
class cmd_runner:

    # --------------------------------------------------------------------
    def __init__(self, ui):
        self.ui = ui
        self.commands = {
            'help' : [self.cmd_help, "Print this help"],
            'quit' : [self.cmd_quit, "Kill the kitten. Using 'force' is an option."],
            'threads' : [self.cmd_threads, "List alive threads"],
            'switch' : [self.cmd_switch, "Summon All Demons of Evil"],
            'stats' : [self.cmd_stats, "Print current statistics (stats [SLEEP] [COUNT])"],
            'set' : [self.cmd_set, "Show/set variables (set PARAMETER VALUE)"],
            'hello' : [self.cmd_hello, "Be polite, say hello"]
        }

    # --------------------------------------------------------------------
    def cmd_help(self, args):
        self.ui.write(" Available commands:")
        for c in sorted(self.commands.keys()):
            self.ui.write("   %-10s : %s " % (c, self.commands[c][1]))

    # --------------------------------------------------------------------
    def cmd_quit(self, args):

        force = False
        try:
            if args[0]:
                if args[0] == "force":
                    force = True
                else:
                    self.ui.write(" I don't know how to quit this way, sorry")
                    return 0
        except:
            pass

        if not force:
            if (cbpx_stats.c_endpoints > 0):
                self.ui.write(" I won't quit with active connections. See stats.")
                return 0
            if (conn_q.qsize() > 0):
                self.ui.write(" I won't quit with connections in queue. See stats.")
                return 0
            self.ui.write(" Exiting...")
        else:
            self.ui.write(" Terminating connections and exiting...")
        return 1

    # --------------------------------------------------------------------
    def cmd_threads(self, args):
        threads = ""
        for t in threading.enumerate():
            threads = "%s %s" % (t.getName(), threads)

        self.ui.write(" Threads: %s" % threads)

    # --------------------------------------------------------------------
    def cmd_switch(self, args):
        active = relay.get_active()
        (ai, ap) = active
        (si, sp) = relay.get_standby()

        self.ui.write(" Starting switch: %s:%i -> %s:%i, timeout: %2.2f s, %i connections buffer\n" % (ai, ap, si, sp, float(params.switch_max_time), int(params.max_queued_conns)))
        l.info("Starting switch: %s:%i -> %s:%i, timeout: %2.2f s, %i connections buffer" % (ai, ap, si, sp, float(params.switch_max_time), int(params.max_queued_conns)))

        # stop relaying connections now
        relay.clear("switch started")

        # set the timer for max switch time
        switch_timer = SwitchTimer(float(params.switch_max_time))
        switch_timer.start()
        switch_start = time.time()

        # print initial stats
        self.print_stats(True, 0)

        l.debug("About to enter switch loop")

        while not relay.isSet():
            try:
                l.debug("Switch loop wait")
                threading.Event().wait(float(params.switch_loop_wait))

                # print stats
                waited = time.time() - switch_start
                self.print_stats(False, waited)
                l.debug("Switch active, waited: %2.2f" % waited)

            except Exception, e:
                l.info("Exception in 'switch' loop: %s" % str(e))

            except KeyboardInterrupt:
                l.warning("Ctrl-c in switch loop, break")
                self.ui.write(" Ctrl-c")
                relay.set("ctrl-c")

        switch_timer.cancel()

        l.debug("Loop done, checking conditions")

        # check what happened and report to user
        reason = relay.get_reason()
        if relay.get_active() == active:
            l.warning("Backend not switched: %s" % reason)
            self.ui.write("\n Switch failed: %s" % reason)
        else:
            l.info("Backend switched: %s" % reason)
            self.ui.write("\n Switch OK!: %s " % reason)


    # --------------------------------------------------------------------
    def print_stats(self, header, switch):
        sw = "[n/a] "
        if switch != -1:
            try:
                sw = "%2.2f" % float(switch)
            except:
                pass
        if header:
            self.ui.write(" SWtime Current backend       active in queue ->cps cps-> q total")
            self.ui.write(" ------ --------------------- ------ -------- ----- ----- -------")
        (ai, ap) = relay.get_active()
        self.ui.write(" %-6s %-21s %6i %8i %5i %5i %7i" % (sw, ai + ":" + str(ap), cbpx_stats.c_endpoints/2, conn_q.qsize(), cbpx_stats.s_qc, cbpx_stats.s_dqc, cbpx_stats.c_qc))

    # --------------------------------------------------------------------
    def cmd_stats(self, args):

        try:
            maxloop = int(args[1])
            if maxloop < 1:
                self.ui.write(" Stats loop count should be >= 1")
                return
        except:
            maxloop = 0

        try:
            sleep = float(args[0])
            if sleep < 0.1:
                self.ui.write(" Stats loop sleep should be >= 0.1")
                return
        except Exception, e:
            sleep = 0

        cnt = 0

        while True:
            if not cnt % 50:
                self.print_stats(True, -1)
            else:
                self.print_stats(False, -1)
            cnt += 1
            try:
                threading.Event().wait(float(args[0]))
                if maxloop and cnt >= maxloop:
                    break
            except Exception, e:
                l.debug("Exception in 'stats' loop: %s, break" % str(e))
                break
            except KeyboardInterrupt:
                l.debug("Ctrl-C in stats loop, break")
                break

    # --------------------------------------------------------------------
    def print_cfg(self):
        if int(params.rc_port) > 0:
            remote = str(params.rc_port)
        else:
            remote = "DISABLED"
        self.ui.write(" Remote control port : %s" % remote)
        self.ui.write(" Listening on        : %i" % params.port)
        self.ui.write(" Active backend      : %s:%i" % (params.active_ip, params.active_port))
        self.ui.write(" Standby backend     : %s:%i" % (params.standby_ip, params.standby_port))
        self.ui.write(" Switch timeout      : %2.2f s" % float(params.switch_max_time))
        self.ui.write(" Max queued conns    : %i" % int(params.max_queued_conns))
        self.ui.write(" Max open conns      : %i" % int(params.max_open_conns))
        self.ui.write(" listen() backlog    : %i" % int(params.listen_backlog))
        self.ui.write(" Network buffer      : %i bytes" % int(params.net_buffer_size))
        self.ui.write(" Switch loop wait    : %2.2f s" % float(params.switch_loop_wait))
        self.ui.write(" Switch script       : %s" % params.switch_script)
        self.ui.write(" Log file            : %s" % params.log_file)
        self.ui.write(" Log level           : %s" % params.log_level)
        self.ui.write("")
        self.ui.write(" Dynamic: %s" % str(params.settable.keys()))

    # --------------------------------------------------------------------
    def cmd_set(self, args):
        # no arguments = prit current settings
        if len(args) == 0:
            self.print_cfg()
            return

        # wrong number of arguments
        if len(args) != 2:
            self.ui.write("Use: 'set PARAMETER VALUE' to change setting")
            return

        # check if parameter is available in configuration
        if not hasattr(params, args[0]):
            self.ui.write(" No such parameter: %s" % args[0])
            return

        # check if parameter can be set
        if args[0] not in params.settable:
            self.ui.write(" Paremeter is not settable: %s" % args[0])
            return

        # check if value type is correct
        try:
            v_test = params.settable[args[0]][0] (args[1])
        except:
            self.ui.write(" VALUE for '%s' must be of %s" % (args[0], params.settable[args[0]][0]))
            return

        # check if value range is correct
        if v_test < params.settable[args[0]][1] or v_test > params.settable[args[0]][2]:
            self.ui.write(" %s must be between %s and %s" % (args[0], str(params.settable[args[0]][1]), str(params.settable[args[0]][2])))
            return

        # everything looks fine, set the parameter
        l.debug("Setting '%s' to '%s'" % (args[0], args[1]))
        try:
            params.__dict__[args[0]] = args[1]
        except Exception, e:
            self.ui.write(" Could not set parameter '%s' to '%s', error: %s" % (args[0], args[1], str(e)))
            l.warning("Could not set parameter '%s' to '%s', error: %s" % (args[0], args[1], str(e)))
            return

        self.ui.write(" Parameter '%s' set to '%s' " % (args[0], params.__dict__[args[0]]))

    # --------------------------------------------------------------------
    def cmd_hello(self, args):
        self.ui.write(" Hello! I'm cbpx %s. And you look great today, I must say." % __version__)

    # --------------------------------------------------------------------
    def process_command(self):
        try:
            line = self.ui.read()
            if not line:
                return
            l.info("Got command: '%s'" % line)
            l_cmd = line.split(" ")[0]
            l_args = line.split(" ")[1:]
            if l_cmd and (l_cmd not in self.commands.keys()):
                self.ui.write(" Unknown command: '%s'" % l_cmd)
                self.ui.finish()
            else:
                res = self.commands[l_cmd][0](l_args)
                self.ui.finish()
                return res
        except KeyboardInterrupt:
            l.debug("Got KeyboardInterrupt, ignoring")
            self.ui.write("")
        except EOFError:
            self.ui.write("")
        except Exception, e:
            l.warning("Exception %s: %s" % (type(e), str(e)))
        return

    # --------------------------------------------------------------------
    def run(self):
        while True:
            if self.process_command():
                break
        self.ui.shutdown()

# vim: tabstop=8 expandtab shiftwidth=4 softtabstop=4


proxy.py
#  Copyright (c) 2011-2012 Jakub Filipowicz <jakubf@gmail.com>
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software
#  Foundation, Inc.,
#  51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

from utils import params, l
from network import cbpx_listener, cbpx_connector, cbpx_transporter
from cmd import cmd_runner
from ui import ui_net, ui_readline
from stats import cbpx_stats

# ------------------------------------------------------------------------
class cbpx:

    # --------------------------------------------------------------------
    def __init__(self):

        # start gathering statistics

        try:
            l.debug("Proxy setting up stats")
            self.stats = cbpx_stats()
        except Exception, e:
            raise RuntimeError("Error setting up stats: %s" % str(e))

        # start transporter

        try:
            l.debug("Proxy setting up Transporter")
            self.transporter = cbpx_transporter()
        except Exception, e:
            raise RuntimeError("Error setting up transporter: %s" % str(e))

        # start connector

        self.connectors = []
        try:
            l.debug("Proxy setting up Connectors")
            # start 5 connectors to make sure connections are processed even if TCP connection to the backend
            # stalls due to lost syn/syn-ack/ack packet
            for i in range(0, 5):
                self.connectors.append(cbpx_connector(params.active_ip, params.active_port, params.standby_ip, params.standby_port, self.transporter))
        except Exception, e:
            raise RuntimeError("Error setting up connector: %s" % str(e))

        # start listener

        try:
            l.debug("Proxy setting up Listener")
            self.listener = cbpx_listener(params.port)
        except Exception, e:
            raise RuntimeError("Error setting up listener: %s" % str(e))

        # setup command interface

        try:
            if int(params.rc_port) > 0:
                l.debug("Proxy setting up command runner with network user interface")
                self.cmd = cmd_runner(ui_net())
            else:
                l.debug("Proxy setting up command runner with readline user interface")
                self.cmd = cmd_runner(ui_readline())
        except Exception, e:
            raise RuntimeError("Error setting up command interface: %s" % str(e))

    # --------------------------------------------------------------------
    def close(self):
        l.debug("Proxy closing listener")
        try: self.listener.close()
        except: pass
        l.debug("Proxy joining listener")
        self.listener.join()

        l.debug("Proxy closing connectors")
        for c in self.connectors:
            try: c.close()
            except: pass
        l.debug("Proxy joining connectors")
        for c in self.connectors:
            c.join()

        l.debug("Proxy closing stats")
        try: self.stats.close()
        except: pass
        l.debug("Proxy joining stats")
        self.stats.join()

        l.debug("Proxy closing transporter")
        try: self.transporter.close()
        except: pass

        l.debug("Proxy killing connections (if any) on close()")
        self.transporter.kill_connections()

        l.debug("Proxy joining transporter")
        self.transporter.join()

    # --------------------------------------------------------------------
    def run(self):
        l.debug("Proxy starting stats")
        self.stats.start()
        l.debug("Proxy starting transporter")
        self.transporter.start()
        l.debug("Proxy starting connectors")
        for c in self.connectors:
            c.start()
        l.debug("Proxy starting listener")
        self.listener.start()
        l.debug("Proxy starting command interface")
        self.cmd.run()
        l.debug("Exiting proxy run() loop")

# vim: tabstop=4 expandtab shiftwidth=4 softtabstop=4


utils.py
#  Copyright (c) 2011-2012 Jakub Filipowicz <jakubf@gmail.com>
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software
#  Foundation, Inc.,
#  51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

import optparse
import logging

l = logging.getLogger()

__version__ = "__cbpx__version__"

# ------------------------------------------------------------------------
# class for storing configuration provided by optparse as well as our own stuff
class my_params:

    settable = {'switch_max_time':[float, 0, 10], 'max_queued_conns':[int, 10, 10000], 'max_open_conns':[int, 10, 65536], 'switch_loop_wait':[float, 0.1, 1], 'net_buffer_size':[int, 1024, 65536]}

    # network:
    port = 0
    rc_port = 0
    active = ''
    standby = ''
    active_ip = ''
    active_port = ''
    standby_ip = ''
    standby_port = ''
    listen_backlog = '512'
    net_buffer_size = '2048'

    # switch:
    max_queued_conns = ''
    max_open_conns = ''
    switch_max_time = ''
    switch_loop_wait = '0.1'
    switch_script = ''

    # logging
    log_file = 'cbpx.log'
    log_level = 'INFO'
    log_format = '%(asctime)-15s %(levelname)-7s [%(threadName)-10s] (%(module)s::%(funcName)s) [L:%(lineno)d] %(message)s'
    log_levels = ['DEBUG', 'INFO', 'WARNING', 'ERROR']

params = my_params()

# ------------------------------------------------------------------------
def parse_cmdline():
    parser = optparse.OptionParser(description='cbpx ' + __version__ + ' : connection buffering proxy', version="%prog " + __version__)
    parser.add_option('-p', '--port', help='port that proxy will listen on', type=int)
    parser.add_option('-a', '--active', help='IP:port pair of active backend (the one we switch from)')
    parser.add_option('-s', '--standby', help='IP:port pair of standby backend (the one we switch to)')
    parser.add_option('-r', '--rc_port', help='port for remote control connections (RC disabled if not specified)', type=int)
    parser.add_option('-t', '--switch_max_time', help='timeout (in seconds) after which switchover fails')
    parser.add_option('-c', '--max_queued_conns', help='queued connections limit, after which switchover fails')
    parser.add_option('-o', '--max_open_conns', help='open connections limit (used for throttling, 0 disables this feature)')

    parser.add_option('-b', '--listen_backlog', help='backlog for listen()')
    parser.add_option('-n', '--net_buffer_size', help='network communication buffer size')
    parser.add_option('-w', '--switch_loop_wait', help='wait in switch loop')
    parser.add_option('-f', '--log_file', help='log file name')
    parser.add_option('-l', '--log_level', help='log level: %s' % str(my_params.log_levels))
    parser.add_option('-x', '--switch_script', help='script to execute before switch completion (optional)')

    (params, args) = parser.parse_args(values=my_params)

    # check for required parameters
    if not params.port:
        raise SyntaxError("-p PORT is required")
    if not params.active:
        raise SyntaxError("-a ACTIVE is required")
    if not params.standby:
        raise SyntaxError("-s STANDBY is required")
    if not params.switch_max_time:
        raise SyntaxError("-t MAX_TIME is required")
    if not params.max_queued_conns:
        raise SyntaxError("-c MAX_QUEUED_CONNS is required")
    if not params.max_open_conns:
        raise SyntaxError("-o MAX_OPEN_CONNS is required")
    if params.log_level not in params.log_levels:
        raise SyntaxError("Log level must be one of: %s ('%s' is wrong)" % (str(params.log_levels), params.log_level))

    # get IP and PORT from what user provided
    try:
        params.active_ip = params.active.split(":")[0]
        params.active_port = int(params.active.split(":")[1])
    except Exception, e:
        raise SyntaxError("specify active backend as IP:PORT, not '%s'\nException: %s" % (params.active, str(e)))

    try:
        params.standby_ip = params.standby.split(":")[0]
        params.standby_port = int(params.standby.split(":")[1])
    except Exception, e:
        raise SyntaxError("specify standby backend as IP:PORT, not '%s'\nException: %s" % (params.standby, str(e)))

# ------------------------------------------------------------------------
def setup_logging():
    logging.basicConfig(format=params.log_format, filename=params.log_file)
    l.setLevel(logging.__dict__["_levelNames"][params.log_level])



# vim: tabstop=8 expandtab shiftwidth=4 softtabstop=4
